src/main/java/de/tum/in/www1/artemis/repository/ProgrammingExerciseRepository.java

```diff
@@ -29,9 +29,19 @@
 @Repository
 public interface ProgrammingExerciseRepository extends JpaRepository<ProgrammingExercise, Long> {
 
-    // Does a max join on the result table for each participation by result id (the newer the result id, the newer the result). This makes sure that we only receive the latest
-    // result for the template and the solution participation if they exist.
-    @Query("select distinct pe from ProgrammingExercise pe left join fetch pe.templateParticipation tp left join fetch pe.solutionParticipation sp left join fetch tp.results as tpr left join fetch sp.results as spr where pe.course.id = :#{#courseId} and (tpr.id = (select max(id) from tp.results) or tpr.id = null) and (spr.id = (select max(id) from sp.results) or spr.id = null)")
+    /**
+     * Does a max join on the result table for each participation by result id (the newer the result id, the newer the result).
+     * This makes sure that we only receive the latest result for the template and the solution participation if they exist.
+     * @param courseId the course the returned programming exercises belong to.
+     * @return all exercises for the given course with only the latest results for solution and template each (if present).
+     */
+    @Query("""
+            SELECT pe FROM ProgrammingExercise pe LEFT JOIN FETCH pe.templateParticipation tp LEFT JOIN FETCH pe.solutionParticipation sp
+            LEFT JOIN FETCH tp.results AS tpr LEFT JOIN FETCH sp.results AS spr
+            WHERE pe.course.id = :#{#courseId}
+                AND (tpr.id = (SELECT MAX(id) FROM tp.results) OR tpr.id IS NULL)
+                AND (spr.id = (SELECT MAX(id) FROM sp.results) OR spr.id IS NULL)
+            """)
```

Comment:
to further increase the readability we often put each join into a new line

```suggestion
    @Query("""
            SELECT pe FROM ProgrammingExercise pe
            LEFT JOIN FETCH pe.templateParticipation tp
            LEFT JOIN FETCH pe.solutionParticipation sp
            LEFT JOIN FETCH tp.results tpr
            LEFT JOIN FETCH sp.results spr
            WHERE pe.course.id = :#{#courseId}
                AND (tpr.id = (SELECT MAX(id) FROM tp.results) OR tpr.id IS NULL)
                AND (spr.id = (SELECT MAX(id) FROM sp.results) OR spr.id IS NULL)
            """)
```